web/pdf_viewer.js

```diff
@@ -2259,13 +2293,56 @@ class PDFViewer {
     if (!this.pdfDocument) {
       return;
     }
-    this.#annotationEditorMode = mode;
-    this.eventBus.dispatch("annotationeditormodechanged", {
-      source: this,
-      mode,
-    });
 
-    this.#annotationEditorUIManager.updateMode(mode, editId, isFromKeyboard);
+    const { eventBus } = this;
+    const updater = () => {
+      if (this.#onPageRenderedCallback) {
+        eventBus._off("pagerendered", this.#onPageRenderedCallback);
+        this.#onPageRenderedCallback = null;
+      }
+      if (this.#switchAnnotationEditorModeTimeoutId !== null) {
+        clearTimeout(this.#switchAnnotationEditorModeTimeoutId);
+        this.#switchAnnotationEditorModeTimeoutId = null;
+      }
+      this.#annotationEditorMode = mode;
+      eventBus.dispatch("annotationeditormodechanged", {
+        source: this,
+        mode,
+      });
+      this.#annotationEditorUIManager.updateMode(mode, editId, isFromKeyboard);
+    };
+
+    if (
+      mode === AnnotationEditorType.NONE ||
+      this.#annotationEditorMode === AnnotationEditorType.NONE
+    ) {
+      const isEditing = mode !== AnnotationEditorType.NONE;
+      if (!isEditing) {
+        this.pdfDocument.annotationStorage.resetModifiedIds();
+      }
+      for (const pageView of this._pages) {
+        pageView.toggleEditingMode(isEditing);
+      }
+      // We must call #switchToEditAnnotationMode unconditionally to ensure that
+      // page is rendered if it's useful or not.
+      const idsToRefresh = this.#switchToEditAnnotationMode();
+      if (isEditing && editId && idsToRefresh) {
+        // We're editing an existing annotation so we must switch to editing
+        // mode when the rendering is done.
+        const { signal } = this.#eventAbortController;
+        this.#onPageRenderedCallback = ({ pageNumber }) => {
+          idsToRefresh.delete(pageNumber);
+          if (idsToRefresh.size === 0) {
+            eventBus._off("pagerendered", this.#onPageRenderedCallback);
+            this.#onPageRenderedCallback = null;
+            this.#switchAnnotationEditorModeTimeoutId = setTimeout(updater, 0);
+          }
+        };
+        eventBus._on("pagerendered", this.#onPageRenderedCallback, { signal });
```

Comment:
What if the following order of events occur:
 - The user switches from editing mode NONE to e.g. INK.
 - The pages are complex enough that re-rendering takes a little while.
 - The user *immediately* switches from INK to e.g. STAMP, before rendering has finished.

It seems, unless I'm misunderstanding things, that in such a case there's nothing that'd remove the *previously* registered "pagerendered" event listener and that one would thus remain pending forever.
Hence, don't we need to essentially run https://github.com/mozilla/pdf.js/pull/18134/files#diff-4238fc7e96d72f4095859850f3d24a9298ec2dc8bd96437214a02dccbd48385cR2299-R2306 *before* registering this event listener here?

Reply:
I think this case is already handled thanks to:
https://github.com/mozilla/pdf.js/pull/18134/files#diff-4238fc7e96d72f4095859850f3d24a9298ec2dc8bd96437214a02dccbd48385cR2315-R2317

We only re-render when we switch to/from NONE. Here in your case we're switch from INK to STAMP.

Reply:
@calixteman Are you sure, since in the case I'm outlining rendering has *not* finished before the INK to STAMP swtich which means that we've still not updated the `this.#annotationEditorMode`-field yet and the `if` you're referencing will thus be true?

Reply:
Yep you're right: the user should first double click on a freetext and then while waiting double click on a ink... and then we're in the case you mentioned: I'm fixing.